GradeServer.c

```diff
@@ -1,146 +1,216 @@
-    pthread_mutex_unlock(&task_list_mutex);
+task_t get_task(void)
+{
+    pthread_mutex_lock(&task_mutex);
+
+    while (num_tasks == 0)
+        pthread_cond_wait(&task_cond, &task_mutex);
+
+    // Get the next task from the list
+    task_t task = tasks[--num_tasks];
+
+    pthread_mutex_unlock(&task_mutex);
+    return task;
 }
 
-int main(int argc, char* argv[]) {
-  if (argc != 2) {
-    fprintf(stderr, "Usage: %s <port>\n", argv[0]);
-    return 1;
-  }
-
-  struct sigaction sa;
-  memset(&sa, 0, sizeof(sa));
-  sa.sa_handler = sigint_handler;
-  sigaction(SIGINT, &sa, NULL);
-
-  // Load the assistants and students from files
-  assistants[MAX_LINE_LENGTH];
-  int num_assistants = loadUsersFromFile(assistants, ASSISTANTS_FILE, ASSISTANT_USER_TYPE);
-  int num_students = loadUsersFromFile(assistants, STUDENTS_FILE, STUDENT_USER_TYPE);
-  int student_grades[num_students] = {0};
-
-  // Print the data from the data structure - TEST PRINT! Remove after checking it works
-  for (int i = 0; i < num_users; i++) {
-     printf("ID: %s, Password: %s\n", assistants[i].id, assistants[i].password);
-  }
-
-  // Init task list and locks
-  task_list_size = 0;
-  pthread_mutex_init(&task_list_mutex, NULL);
-  pthread_cond_init(&task_list_cond, NULL);
-
-  // Get the port number from the command line argument
-  int port = atoi(argv[1]);
-
-  // Create a socket
-  sockfd = socket(AF_INET, SOCK_STREAM, 0);
-  if (sockfd < 0) {
-    perror("Error opening socket");
-    return 1;
-  }
-
-  // Set up the server address
-  struct sockaddr_in serv_addr;
-  memset(&serv_addr, 0, sizeof(serv_addr));
-  serv_addr.sin_family = AF_INET;
-  serv_addr.sin_addr.s_addr = INADDR_ANY;
-  serv_addr.sin_port = htons(port);
-
-  // Bind the socket to the server address
-  if (bind(sockfd, (struct sockaddr*) &serv_addr, sizeof(serv_addr)) < 0) {
-    perror("Error binding socket");
-    return 1;
-  }
-
-  // Listen for incoming connections
-  listen(sockfd, 5);
-
-  // Create a pool of worker threads
-  pthread_t threads[NUM_THREADS];
-  for (int i = 0; i < NUM_THREADS; i++) {
-    if (pthread_create(&threads[i], NULL, handle_connection, NULL) != 0) {
-      perror("Error creating thread");
-      return 1;
-    }
-  }
-
-  // Infinite loop - accept user's connection and adding task for every new connection
-  while (1) {
-    // Accept a connection
-    struct sockaddr_in cli_addr;
-    socklen_t cli_addr_len = sizeof(cli_addr);
-    int connfd = accept(sockfd, (struct sockaddr*) &cli_addr, &cli_addr_len);
-    if (connfd < 0) {
-      perror("Error accepting connection");
-      return 1;
+void delete_task(task_t task)
+{
+    pthread_mutex_lock(&task_mutex);
+
+    for (int i = 0; i < num_tasks; i++)
+    {
+        if (strcmp(tasks[i].message, task.message) == 0)
+        {
+            // Shift all tasks after the specified index one position to the left
+            for (int j = i; j < num_tasks - 1; j++)
+                tasks[j] = tasks[j + 1];
```

Comment:
If we move to the "left", shouldn't it be `tasks[j] = tasks[j - 1];` instead of +?

Reply:
First, it's the chat's function 😅
Second, I think by "shifting to the left" it means to "move the pointer to the right", so it's j=j+1